torch/nested/_internal/ops.py

```diff
@@ -0,0 +1,156 @@
+        is_optional = arg_type.endswith("?")
+        normalized_arg_type = arg_type[:-1] if is_optional else arg_type
+        if not normalized_arg_type in arg_type_check_fns.keys():
+            raise AssertionError(f"Unknown arg type: {normalized_arg_type}")
+
+        if i >= len(args):
+            if not is_optional:
+                raise ValueError(f"NestedTensor {func.__name__}({schema_str}) "
+                                 f"missing required argument: {name}")
+            continue
+
+        if not arg_type_check_fns[normalized_arg_type](args[i]):
+            raise ValueError(
+                f"NestedTensor {func.__name__}({schema_str}): {name} should be of "
+                f"type {arg_type}, but got: {type(args[i])}")
+
+def check_ragged_dim_same(
+        func,
+        a: NestedTensor,
+        a_name: str,
+        b: NestedTensor,
+        b_name: str) -> None:
+    # Calling into .shape here
+    assert len(a._size) == 3, "NestedTensor must be [B, *, D]"
+    if a._size[1] != b._size[1]:
+        raise RuntimeError(
+            f"NestedTensor {func.__name__}: expected {a_name} and {b_name} to have the "
+            "same exact offsets tensor.")
+
+def register_func(tables, aten_ops, schema_str):
+    if not isinstance(aten_ops, list):
+        aten_ops = [aten_ops]
+    if not isinstance(tables, list):
+        tables = [tables]
+
+    def wrapper(func):
+        for aten_op in aten_ops:
+            def get_inner(aten_op):
+                def inner(*args, **kwargs):
+                    check_schema(schema_str, func, *args, **kwargs)
+                    return func(aten_op, *args, **kwargs)
+                return inner
+
+            for table in tables:
+                table[aten_op] = get_inner(aten_op)
+    return wrapper
+
+register_jagged_func = functools.partial(register_func, JAGGED_OPS_TABLE)
+
+def lookup_jagged(func, *args, **kwargs) -> Optional[Callable]:
+    if torch.Tag.pointwise in func.tags:
+        # Assume there aren't additional tensors that aren't the "unary/binary" args
+        num_tensor_args = sum([isinstance(x, torch.Tensor) for x in args])
+        if num_tensor_args == 1:
+            return functools.partial(jagged_unary_pointwise, func)
+        elif num_tensor_args == 2:
+            check_schema("lhs: jt, rhs: jt", func, *args, **kwargs)
+            return functools.partial(jagged_binary_pointwise, func)
+        else:
+            return None
+    return JAGGED_OPS_TABLE.get(func, None)
+
+def extract_kwargs(arg):
+    kwargs = {
+        "offsets": arg.offsets,
+        "nb_tensors": arg.nb_tensors,
+    }
+    return kwargs
+
+def jagged_unary_pointwise(func, *args, **kwargs):
+    return NestedTensor(func(args[0].buffer, **kwargs), **extract_kwargs(args[0]))
+
+def jagged_binary_pointwise(func, *args, **kwargs):
+    check_ragged_dim_same(func, args[0], "lhs", args[1], "rhs")
+    return NestedTensor(func(args[0].buffer, args[1].buffer, **kwargs), **extract_kwargs(args[0]))
```

Comment:
it may not be correct to unconditionally take the offsets of the first arg

Reply:
yeah, haven't thought about the broadcasting case yet, but maybe okay for now

Reply:
as long as we guard on same ragged structure yeah